### ISSUE TITLE
Support for let's encrypt

### ISSUE THREAD

**Issue by tonioo** [modoboa/modoboa-installer#50]
_From @tisba37 on April 2, 2016 10:51_

Hello,

It's prossible to add a support for let's encrypt ?

_Copied from original issue: tonioo/modoboa#856_

**Comment by tonioo** [comment]
_From @hadifarnoud on April 9, 2016 14:41_

native support would be awesome. Mailinabox did something in that effect. can we re-use some of that code?

**Comment by tonioo** [comment]
Maybe. Once again, don't hesitate to contribute.

**Comment by tonioo** [comment]
_From @spinza on May 9, 2016 23:38_

I'm running with letsencrypt certificates.  Not that difficult to setup.

**Comment by tonioo** [comment]
That's something the installer could do pretty easily.

**Comment by tonioo** [comment]
_From @almereyda on May 10, 2016 15:47_

Before installing, I have been using the following procedure to generate certificates.

---

Create a letsencrypt certificate with https://github.com/lukas2511/letsencrypt.sh and Python SimpleHTTPServer, assuming you work as `root`.

1. Clone into `/root/letsencrypt.sh`: `git clone https://github.com/lukas2511/letsencrypt.sh`
Create the following files within that folder:
    * `domains.txt`
    ```
    FQDN of your box
    ```
    * `config.sh`
    ```
    WELLKNOWN="/root/webroot/.well-known/acme-challenge"
    CONTACT_EMAIL="hostmaster@FQDN"
    ```

2. Create your HTTP webroot `mkdir -p /root/webroot/.well-known/acme-challenge`
    * Run an HTTP server from there `cd /root/webroot && python -m SimpleHTTPServer 80`

3. Run `./letsencrypt.sh -e` to verify your settings and `./letsencrypt.sh -c` to (re-)generate the certificate(s).

---

Then the `installer.cfg` from Modoboa's installer would receive the following lines:

```
[general]
tls_key_file = /root/letsencrypt.sh/certs/FQDN/privkey.pem
tls_cert_file = /root/letsencrypt.sh/certs/FQDN/fullchain.pem
```

**Comment by almereyda** [comment]
The tricky part will be the automatic regeneration by making sure to run `./letsencrypt.sh -c` every once in a while and reloading `postfix` as well as `nginx` when an update occurs.

**Comment by almereyda** [comment]
Also see #55

**Comment by tonioo** [comment]
@almereyda @hadifarnoud Anyone to test this new feature please ? :-)

**Comment by almereyda** [comment]
Yes, I can do so. Was thinking about the refresh-restart (postfix, web) logic today, as I said on June 7 above.
Is automatic refreshment included? `ssl.py` seems to be equipped for that. Let's see how the installer can build a working instance from there.

**Comment by almereyda** [comment]
Hi @tonioo me and @gandhiano have just tested the feature branch on our production machine.

It happened that we had to remove `webroot, ` from https://github.com/modoboa/modoboa-installer/commit/7c22bbe5f08ba7e5dafd6e28339934e0828f77db#diff-c226f84131b50059e044a64c41fe782cR78, as the `certbot` is run as standalone.

Then `./run.py email.ecobytes.net --debug` tells us few errors about the services startup, but they are running either way.

```
Synchronizing state for nginx.service with sysvinit using update-rc.d...
Executing /usr/sbin/update-rc.d nginx defaults
Executing /usr/sbin/update-rc.d nginx enable
● nginx.service - A high performance web server and a reverse proxy server
   Loaded: loaded (/lib/systemd/system/nginx.service; enabled)
   Active: failed (Result: exit-code) since Fri 2016-10-07 15:17:10 BST; 4min 54s ago
 Main PID: 704 (code=exited, status=0/SUCCESS)

Oct 07 15:17:10 email nginx[16056]: nginx: [emerg] BIO_new_file("/etc/letsencrypt/live/email.ecobytes.net/fullchain.p...h file)
Oct 07 15:17:10 email nginx[16056]: nginx: configuration file /etc/nginx/nginx.conf test failed
Oct 07 15:17:10 email systemd[1]: nginx.service: control process exited, code=exited status=1
Oct 07 15:17:10 email systemd[1]: Failed to start A high performance web server and a reverse proxy server.
Oct 07 15:17:10 email systemd[1]: Unit nginx.service entered failed state.
Hint: Some lines were ellipsized, use -l to show in full.
```

At the first try this happened, as `certbot` didn't manage to create the certificates, which nginx expected in its configuration. `nginx -t` failed also.

```
● dovecot.service - Dovecot IMAP/POP3 email server
   Loaded: loaded (/lib/systemd/system/dovecot.service; enabled)
   Active: failed (Result: exit-code) since Fri 2016-10-07 15:17:19 BST; 4min 53s ago
 Main PID: 16731 (code=exited, status=89)

Oct 07 15:17:19 email dovecot[16731]: doveconf: Fatal: Error in configuration file /etc/dovecot/conf.d/10-ssl.conf lin...ectory
Oct 07 15:17:19 email systemd[1]: dovecot.service: main process exited, code=exited, status=89/n/a
Oct 07 15:17:19 email systemd[1]: Unit dovecot.service entered failed state.
Hint: Some lines were ellipsized, use -l to show in full.
Congratulations! You can enjoy Modoboa at https://email.ecobytes.net (admin:password)
```

As nginx was not running, the congratulations were also wrong, but we tried it again. We still got the same error messages, but the services are running as expected:

```
root@email:~/modoboa-installer# systemctl status nginx dovecot postfix
● nginx.service - A high performance web server and a reverse proxy server
   Loaded: loaded (/lib/systemd/system/nginx.service; enabled)
   Active: active (running) since Fri 2016-10-07 15:22:05 BST; 7min ago
 Main PID: 17485 (nginx)
   CGroup: /system.slice/nginx.service
           ├─17485 nginx: master process /usr/sbin/nginx -g daemon on; master_process on;
           ├─17486 nginx: worker process
           ├─17487 nginx: worker process
           ├─17488 nginx: worker process
           └─17490 nginx: worker process

● dovecot.service - Dovecot IMAP/POP3 email server
   Loaded: loaded (/lib/systemd/system/dovecot.service; enabled)
   Active: active (running) since Fri 2016-10-07 15:22:13 BST; 7min ago
 Main PID: 18171 (dovecot)
   CGroup: /system.slice/dovecot.service
           ├─18171 /usr/sbin/dovecot -F
           ├─18173 dovecot/anvil
           ├─18174 dovecot/log
           ├─18176 dovecot/config
           ├─18223 dovecot/auth
           └─18472 dovecot/ssl-params

● postfix.service - LSB: Postfix Mail Transport Agent
   Loaded: loaded (/etc/init.d/postfix)
  Drop-In: /run/systemd/generator/postfix.service.d
           └─50-postfix-$mail-transport-agent.conf
   Active: active (running) since Fri 2016-10-07 15:22:07 BST; 7min ago
   CGroup: /system.slice/postfix.service
           ├─17683 /usr/lib/postfix/master
           ├─17687 pickup -l -t unix -u -c
           ├─17688 qmgr -l -t unix -u
           ├─18376 trivial-rewrite -n rewrite -t unix -u -c
           ├─18423 lmtp -t unix -u -c
           ├─18464 cleanup -z -t unix -u -c
           ├─18477 smtpd -n submission -t inet -u -c -o stress= -o syslog_name=postfix/submission -o smtpd_tls_security_level=encrypt -o smtpd_sasl_auth_enable=yes -o smtpd_reject_unlisted_recipient=no -o smtpd_client_restrictions=permit_sasl_authenti...
           ├─18478 proxymap -t unix -u
           ├─18479 tlsmgr -l -t unix -u -c
           ├─18480 anvil -l -t unix -u -c
           ├─18490 smtpd -n 127.0.0.1:10025 -t inet -u -o content_filter= -o smtpd_authorized_xforward_hosts=127.0.0.0/8 -o smtpd_delay_reject=no -o smtpd_client_restrictions=permit_mynetworks,reject -o smtpd_helo_restrictions= -o smtpd_sender_restric...
           └─18491 smtp -t unix -u -c

Hint: Some lines were ellipsized, use -l to show in full.
```

We didn't start them, but we're happy it worked. Not *like a charm* as people like to say, but getting better.

**Comment by tonioo** [comment]
@almereyda Thanks for your feedback !

So basically, all those errors were caused because of the webroot error at certificate generation ?

**Comment by almereyda** [comment]
First, yes. Second, I don't know.

**Comment by vkarpyuk** [comment]
Hi Guys.

I have installed modoboa server a couple of months ago. Honestly, I have been using it since and I love the performance and everything. However, I found that each time I send an email to a new mail, the receiver receives the mail in spam. I have tested it with Google, Exchange, and others. All of the receivers received my mail in the spam folder. Once they move it to the "income folder" it won't send to spam again. That's been a problem for me because one of my clients is a big organization and sometimes I have to copy or send mails to a new user. I wanted to check what can I do to improve the acceptance and the security so my mails are not directed to spam.

Thanks in advance!

Viktor Karpyuk
